CMake: Refactor UNITTESTS CMake

rajkan01 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
- All the subs cpp files are grouped into their respective libraries like `mbed-stubs-connectivity`, `mbed-stubs-drivers`, `mbed-stubs-events`, `mbed-stubs-hal`, `mbed-stubs-platform`, `mbed-stubs-rtos`, `mbed-stubs-storage` so all libraries(drivers, platform etc.,) unittest CMakeLists.txt target link aginst to specifically required stub lib for their test or use mbed-stubs libraries which intern target linked to all stub libraries
- All the Mbed OS headers are also grouped into header libraries  like `mbed-headers-base`, `mbed-headers-platform`, `mbed-headers-connectivity`, `mbed-headers-storage`, `mbed-headers-drivers`, `mbed-headers-hal`, `mbed-headers-events`, `mbed-headers-rtos` so all libraries (drivers, platform etc.,) unit test CMakeLists.txt target link aginst to specifically required header interface lib or use mbed-headers libraries which intern target linked with all specific headers libraries
- Add the command line argument BUILD_TESTING to ON/OFF Mbed OS unittest build



<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
As part of this PR, we migrated to use the latest version CMake 3.19 or above version to build unit tests and the user must update their CMake to the latest version stated above
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
| Previous unittest build  | New way of unittest build |
| ------------- | ------------- |
| The unittest CMake build discover the library unittest by searching the unittest.cmake file across Mbed OS and build | Every library CMakeLists.txt ( For ex: mbed-os/platform/CMakeLists.txt) will also add the UNITTESTS directory into the build based on NOT CMAKE_CROSSCOMPILING guard and normally all library directory gets added from mbed os root. |
| Every library unittest unittest.cmake used to set includes, sources, test sources, test flags using unittest-includes, unittest-sources, unittest-test-sources, unittest-test-flags CMake variable | Every library unittest has its CMakeLists.txt file a kind of application build CMake where it calls CMake standard calls target_sources, target_link_libraries, add_executable etc., |

_Note:_
Refer to official Mbed OS latest `debug-test/unit-testing`  documentation `Example section` which explains how to write a unittest for `example class` in a new way 
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->
https://github.com/ARMmbed/mbed-os-5-docs/pull/1427
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->
@hugueskamba @0xc0170 @evedon @LDong-Arm @pan- @paul-szczepanek-arm @jamesbeyond @rwalton-arm @Patater @donatieng 
<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

paul-szczepanek-arm committed:
CMake: Update target_h stub headers

- Avoid using cmsis headers as it brings more redundant dependency into unittest build and for unittest build, all the required OS flags can provide from mbed rtos types header so replaced cmsis header with mbed os types header

rajkan01 committed:
CMake: Refactor storage blockdevice and SFDP unittests cmake

- Add CMake configuration file
- Fix the header inclusion in blockdevice unittests

rajkan01 committed:
CMake: Refactor storage TDBstore and FileSystemStore unittest cmake

- Add CMake configuration file in TDBstore and FileSystemStore unittest.

rajkan01 committed:
CMake: Update build tools python modules to pick refactored way of unittest CMake

- Update mbed_unittest python module to pick unittest from mbed os root
CMake configuration file
- Add BUILD_TESTING flag ON in test python module

rajkan01 committed:
CMake: Refactor lorawan unittest cmake

- Add CMakeLists.txt file to all lorawan test suite

rajkan01 committed:
CMake: Refactor netsocket unittest CMake

- Add Cmake configuration file
- Fix the header inclusion in DTLSSocketWrapper and TLSSocketWrapper test

paul-szczepanek-arm commented on UNITTESTS/CMakeLists.txt:
```diff
@@ -1,66 +1,35 @@
-cmake_minimum_required(VERSION 3.0.2)
+# Copyright (c) 2021 ARM Limited. All rights reserved.
```
indentation in this file is a bit off in places, maybe worth just going through and cleaning it up

paul-szczepanek-arm reviewed (dismissed):
Herculean task (akin to cleaning up the Augean stables).

paul-szczepanek-arm commented on storage/kvstore/tdbstore/tests/UNITTESTS/TDBStore/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+
+add_test(NAME "${TEST_NAME}" COMMAND ${TEST_NAME})
+
+set_tests_properties(${TEST_NAME} PROPERTIES LABELS "storage-tdbstore")
```
I would lump all the storage tests into a single `storage` label.

paul-szczepanek-arm reviewed (dismissed):
It all looks great to me. Still needs to update https://github.com/ARMmbed/mbed-os/blob/feature_unittest_refactor/UNITTESTS/README.md.

paul-szczepanek-arm commented on UNITTESTS/fakes/ble/source/GattServerImpl.h:
```diff
@@ -0,0 +1,137 @@
+#define BLE_GATTSERVERSTUB_H
+
+#include "ble/GattServer.h"
+#include "FEATURE_BLE/source/generic/GattServerEvents.h"
```
sorry, this needs:
https://github.com/ARMmbed/mbed-os/pull/14474 or just merge the two suggestions below.

```suggestion
#include "generic/GattServerEvents.h"
```

paul-szczepanek-arm commented on UNITTESTS/fakes/ble/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+    PUBLIC
+        ${MBED_PATH}/UNITTESTS/fakes/ble/
+        ${MBED_PATH}/connectivity/FEATURE_BLE/include
+        ${MBED_PATH}/connectivity/FEATURE_BLE/include/ble
```
```suggestion
        ${MBED_PATH}/connectivity/FEATURE_BLE/include/ble
        ${MBED_PATH}/connectivity/FEATURE_BLE/source
```

Patater commented on .gitignore:
```diff
@@ -106,3 +106,5 @@ platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/TARGE
 CMakeCache.txt
 cmake_install.cmake
 CMakeFiles/
+cmake_build
```
```diff
- cmake_build
+ cmake_build/
```

Patater commented on CMakeLists.txt:
```diff
@@ -5,6 +5,24 @@
 
 cmake_minimum_required(VERSION 3.19.0 FATAL_ERROR)
 
+option(MBED_BUILD_UNITTESTS "Run unit tests only." OFF)
```
Could we use the standard `BUILD_TESTING` instead?

```diff
- option(MBED_BUILD_UNITTESTS "Run unit tests only." OFF)
+ option(BUILD_TESTING "Build the tests." OFF)
```

paul-szczepanek-arm replied:
I think that's the define that is set my `mbed test --unittest`

Patater replied:
We might want to handle it for backwards compat reasons then. BUILD_TESTING is the standard CMake one, so we should use that. We can have the top-level CMakeLists.txt see if `MBED_BUILD_UNITTESTS` is present and then set `BUILD_TESTING` for you accordingly.

rajkan01 replied:
`include(CTest)` will automatically set `BUILD_TESTING` to `ON`. We discussed this point here https://github.com/ARMmbed/mbed-os-5-docs/pull/1427#discussion_r605601593 and https://github.com/ARMmbed/mbed-os-5-docs/pull/1427#discussion_r605632860, so going further we will add `project(MbedOS)` in `mbed-os/CMakeList.txt` and use this condition  `if(CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)` to include CTest

Patater commented on CMakeLists.txt:
```diff
@@ -5,6 +5,24 @@
+option(MBED_BUILD_UNITTESTS "Run unit tests only." OFF)
+
+if(MBED_BUILD_UNITTESTS)
+    project(mbed-unittests)
```
Why do we need to make another project?

rajkan01 replied:
With these changes https://github.com/ARMmbed/mbed-os/pull/14426#discussion_r607693067 we can avoid

Patater commented on CMakeLists.txt:
```diff
@@ -5,6 +5,24 @@
+
+if(MBED_BUILD_UNITTESTS)
+    project(mbed-unittests)
+    enable_testing()
```
If we `include(CTest)`, `enable_testing()` will be called for us. Are we wanting to avoid `include(CTest)` for some reason?

rajkan01 replied:
Removed the enable_testing() API call with the help of include(CTest)

Patater commented on CMakeLists.txt:
```diff
@@ -5,6 +5,24 @@
+    enable_testing()
+    
+    add_subdirectory(UNITTESTS)
+    add_subdirectory(${MBED_PATH}/drivers/tests/UNITTESTS)
```
Is using `${MBED_PATH}` necessary? Aren't the paths here relative to the current source directory, which is already `mbed-os`?

rajkan01 replied:
Removed ${MBED_PATH}

Patater commented on CMakeLists.txt:
```diff
@@ -5,6 +5,24 @@
+    enable_testing()
+    
+    add_subdirectory(UNITTESTS)
+    add_subdirectory(${MBED_PATH}/drivers/tests/UNITTESTS)
```
Why do we add such deep folders here? Wouldn't adding the drivers subdirectory also add the tests subdirectory via the drivers/CMakeLists.txt?

rajkan01 replied:
Modified to use actual library CMake to add unit test sub directory

Patater commented on CMakeLists.txt:
```diff
@@ -5,6 +5,24 @@
+    add_subdirectory(${MBED_PATH}/storage/kvstore/filesystemstore/tests/UNITTESTS)
+    add_subdirectory(${MBED_PATH}/platform/tests/UNITTESTS)   
+    add_subdirectory(${MBED_PATH}/events/tests/UNITTESTS)
+else()
```
Why the else? Shouldn't we always add all targets and let CMake decide what to build based on the requested target?

rajkan01 replied:
We need to keep `else()` to avoid all the target configuration (we are not  generating mbed_build.cmake which is based on target config in targets.json) check like validating target support std vs small library and profile selection check etc.,

Patater replied:
That seems like a bug in the target validating CMake code. It shouldn't bother checking things that don't matter for the chosen target. Are we checking `CMAKE_CROSSCOMPILING` or `CMAKE_SYSTEM_NAME` to ensure that we don't try to validate things that are invalid or irrelevant for the target platform?

0xc0170 replied:
We do not but we shall, to check `CMAKE_CROSSCOMPILING`.

rajkan01 replied:
Used CMAKE_CROSSCOMPILING

Patater commented on UNITTESTS/CMakeLists.txt:
```diff
@@ -1,240 +1,68 @@
+include(FetchContent)
+
+# Download and unpack googletest
+if(NOT GTEST_FOUND)
```
What could set `GTEST_FOUND`?

rajkan01 replied:
I have wrongly mixed this variable with `fetchcontent_xyz` calls as this suppose to be used with the find_pacakage command so I will remove this condition check.
`FetchContent_MakeAvailable` is already taken care of to avoid the second time download when we called the same command again

Patater commented on UNITTESTS/CMakeLists.txt:
```diff
@@ -1,240 +1,68 @@
+add_subdirectory(stubs)
+add_subdirectory(fakes)
 
+macro( mbed_add_all_subdirectories subdirectories_list )
```
Is our CMake style to have or not have whitespace after the opening `(`?

rajkan01 replied:
This is redundant white space, I will remove

Patater commented on UNITTESTS/README.md:
```diff
@@ -12,11 +12,10 @@ Please install the following dependencies to use Mbed OS unit testing:
 * GNU toolchains.
    * GCC 6 or later. We recommend you use MinGW-W64 on Windows, but any Windows port of the above GCC versions works. Default compilers can be used on Mac OS instead of GCC to shorten build times, but code coverage results can differ.
-* CMake 3.0 or newer.
+* CMake 3.19.0 or newer.
```
Do we want to repeat dependencies here? It makes it hard to see what additional dependencies unit tests add to the ones Mbed OS already requires. It also makes it harder to keep the dependencies list between the Mbed OS doc and the unit test doc synchronized.

rajkan01 replied:
I think we can remove this README.md as it is duplicate content of https://os.mbed.com/docs/mbed-os/v6.9/debug-test/unit-testing.html documentation

Patater commented on UNITTESTS/README.md:
```diff
@@ -63,23 +62,23 @@ In a terminal window:
+All the stub libraries and header libraries are defined under `UNITTESTS/stubs/` directory.
 
-#### Test names
+Libraries for fakes are under `UNITTESTS/stubs/` directory. These provide mock implementations that are meant to replace the stub version that do nothing. Usually these will replace the header files as well as the source files and cannot be used together with their stub equivalents.
```
Could we put the fakes in a ``UNITTESTS/fakes/` directory instead?

paul-szczepanek-arm replied:
Fakes are indeed in UNITTESTS/fakes I think it's the docs that are in error here
```suggestion
Libraries for fakes are under `UNITTESTS/fakes/` directory. These provide mock implementations that are meant to replace the stub version that do nothing. Usually these will replace the header files as well as the source files and cannot be used together with their stub equivalents.
```

rajkan01 replied:
same comment https://github.com/ARMmbed/mbed-os/pull/14426#discussion_r607936486

Patater commented on UNITTESTS/README.md:
```diff
@@ -63,23 +62,23 @@ In a terminal window:
+All the stub libraries and header libraries are defined under `UNITTESTS/stubs/` directory.
 
-#### Test names
+Libraries for fakes are under `UNITTESTS/stubs/` directory. These provide mock implementations that are meant to replace the stub version that do nothing. Usually these will replace the header files as well as the source files and cannot be used together with their stub equivalents.
```
```diff
- the stub version that do nothing
+ the stub implementations which do nothing
```

rajkan01 replied:
same comment https://github.com/ARMmbed/mbed-os/pull/14426#discussion_r607936486

Patater commented on UNITTESTS/README.md:
```diff
@@ -63,23 +62,23 @@ In a terminal window:
+Registering unit tests to run happens automatically, and the test runner handles registration. However, test suites do not automatically build. 
 
-Mbed CLI supports unit tests through the `mbed test --unittests` command. For information on using Mbed CLI, please see the [CLI documentation](https://os.mbed.com/docs/mbed-os/latest/tools/developing-mbed-cli.html).
+For the build system to build a unit test, pass the `ON` or `OFF` value to `MBED_BUILD_UNITTESTS` in the command-line to include all the unit test suite directories into the build.
```
Could we use the standard `BUILD_TESTING` instead?

rajkan01 replied:
same comment https://github.com/ARMmbed/mbed-os/pull/14426#discussion_r607936486, we will remove this readme

Patater commented on UNITTESTS/README.md:
```diff
@@ -128,76 +121,40 @@ With the following steps, you can write a simple unit test. This example creates
-    #include "OtherClass.h"
-
+    #include "MyClass.h"
+    
```
Hmm. I remember this trailing whitespace. I feel this doc is copy-pasted in mbed-os-5-docs via https://github.com/ARMmbed/mbed-os-5-docs/pull/1427

Wouldn't it be easier to maintain and keep the documentation up to date if we only had one copy of it?

rajkan01 replied:
yes we will remove this readme as duplicate content of https://os.mbed.com/docs/mbed-os/v6.9/debug-test/unit-testing.html

Patater commented on UNITTESTS/fakes/ble/BLE.cpp:
```diff
@@ -0,0 +1,163 @@
+
+}
+
+}
```
GitHub shows no new-line at the end of this file. Could you add one please?
<img width="247" alt="Screenshot 2021-03-31 at 15 07 39" src="https://user-images.githubusercontent.com/410816/113157607-f583ae80-9232-11eb-8b85-b33c3449851a.png">

Patater commented on UNITTESTS/fakes/ble/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+
+target_include_directories(mbed-fakes-ble
+    PUBLIC
+        ${MBED_PATH}/UNITTESTS/fakes/ble/
```
Why do we need `${MBED_PATH}` here? Can we think of a way to avoid it? This would help our modules be more modular and reusable outside an Mbed context.

rajkan01 replied:
BLE changes are moved into separate PR.

Patater commented on connectivity/FEATURE_BLE/include/ble/common/BLERoles.h:
```diff
@@ -19,6 +19,49 @@
 
+#if !defined(BLE_ROLE_OBSERVER)
+#define BLE_ROLE_OBSERVER 1
+#endif
```
Is editing BLERoles important in a unit test refactoring PR?

paul-szczepanek-arm replied:
it's required for fakes to work

Patater replied:
Please explain the reasoning in the commit message (not title) of the commit that makes this change.

rajkan01 replied:
BLE changes are moved into separate PR.

Patater commented on connectivity/cellular/tests/UNITTESTS/framework/AT/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+    at_cellularstack
+)
+
+mbed_add_all_subdirectories("${MBED_ADD_UNITTEST_SUBDIRS}")
```
What's wrong with just `add_subdirectory()`?

rajkan01 replied:
Instead of repeatedly calling add_subdirectory we added a macro that receives the list of the directories and adds all directories in the list

rajkan01 replied:
Removed the `mbed_add_all_subdirectories` macro and called directly the  `add_subdirectory` call

Patater commented on connectivity/cellular/tests/UNITTESTS/framework/device/cellularstatemachine/cellularstatemachinetest.cpp:
```diff
@@ -23,7 +23,7 @@
 #include "myCellularDevice.h"
 #include "Thread_stub.h"
-#include "cmsis_os2.h"
+#include "mbed_rtos_types.h"
```
Improvements to cellular unit tests probably should be in a separate PR, or at least well-organized separate commits. I didn't review commit-by-commit in this PR, as it looked like we assumed we wanted to squash all the commits together and then merge.

paul-szczepanek-arm replied:
it's a fix required for the unittest to compile, I think it's a legitimate change to be part of this PR

Patater replied:
Let's go for separate commits then, as it isn't obvious why these changes have to be made, and a commit message could help explain why.

rajkan01 replied:
I have put these changes into a separate commit

Patater commented on connectivity/lorawan/tests/UNITTESTS/features/lorawan/loraphycn779/CMakeLists.txt:
```diff
@@ -0,0 +1,31 @@
+
+target_sources(${TEST_NAME}
+    PRIVATE
+        ${MBED_PATH}/connectivity/lorawan/lorastack/phy/LoRaPHYCN779.cpp
```
Can we think of a way to avoid `${MBED_PATH}`? Could we link against a library target that includes this `cpp` instead?

paul-szczepanek-arm replied:
it's a tradeoff, if you want to use library as dependency then you have to break up the module into tiny libraries since if you include the whole module then it will pull in too many dependencies, most of our modules are not self contained

Patater replied:
Yeah, it's a legit strategy to include just the implementation you are testing when unit testing. Listing the source file or two you are testing, rather than the library, give you an opportunity to replace that file's cpp or header dependencies with stubs, mocks, fakes, and so forth. We'd generally expect a unit test suite to include one real file, as is done here. If we are including 10s of real cpp files (we don't do that here, but do in other places in this PR), then that expresses a sort of need for a library, if not a need to rewrite the tests to be less integration or acceptance test and more unit test (i.e. rewrite the test to test a specific unit in isolation).

> most of our modules are not self contained

We should work on that separately.

For this PR, I suspect it is possible to avoid using `$MBED_PATH}`, even without having to link against a library target.

rajkan01 replied:
In this case if we want refer `LoRaPHYCN779.cpp` then we have to do `../../../../../` change directory it it better to refer with ${mbed-os_SOURCE_DIR}/connectivity/lorawan/lorastack/phy/LoRaPHYCN779.cpp" to give better visibility of actual source relevant path

Patater commented on connectivity/netsocket/tests/UNITTESTS/netsocket/DTLSSocket/CMakeLists.txt:
```diff
@@ -0,0 +1,47 @@
+
+target_sources(${TEST_NAME}
+    PRIVATE
+        ${MBED_PATH}/connectivity/netsocket/source/SocketAddress.cpp
```
Would we link against an `mbed-netsocket` target instead?

rajkan01 replied:
IMO proposed changes (link against `mbed-netsocket`) need to address as part of unit test improvement by the connectivity team, once these base PR changes are available into master

Patater commented on connectivity/netsocket/tests/UNITTESTS/netsocket/DTLSSocket/test_DTLSSocket.cpp:
```diff
@@ -62,5 +62,6 @@ TEST_F(TestDTLSSocket, connect)
 
     stack.return_value = NSAPI_ERROR_OK;
     SocketAddress a("127.0.0.1", 1024);
+    stack.return_socketAddress = a;
```
In commit message "CMake: Fix unittest bug and header inclusion", could you provide more context as to why this change is needed? Copying and pasting an error message into the commit message would also be helpful.

rajkan01 replied:
I have put this change into a separate PR

Patater commented on storage/blockdevice/tests/UNITTESTS/blockdevice/ChainingBlockDevice/test_ChainingBlockDevice.cpp:
```diff
@@ -96,7 +96,7 @@ TEST_F(ChainingBlockModuleTest, init)
 //    EXPECT_EQ(b.deinit(), BD_ERROR_OK);
     EXPECT_EQ(b.sync(), BD_ERROR_DEVICE_ERROR);
-    EXPECT_EQ(b.get_type(), "CHAINING");
+    EXPECT_STREQ(b.get_type(), "CHAINING");
```
In commit "CMake: Fix stubs header inclusion", could you explain what motivates the changes? If `EXPECT_STREQ` doesn't help fix header inclusion, it probably needs a separate commit with separate justification.

rajkan01 replied:
I have put this change into separate commit

Patater commented on storage/blockdevice/tests/UNITTESTS/blockdevice/HeapBlockDevice/CMakeLists.txt:
```diff
@@ -0,0 +1,67 @@
+
+target_sources(${TEST_NAME}
+    PRIVATE
+        ${MBED_PATH}/storage/blockdevice/source/HeapBlockDevice.cpp
```
If we link against a library that provides a heap block device instead of listing the sources here, we'll help move Mbed to be more modular. The config you've added in `target_compile_definitions()` in this file would move to the library target and wouldn't need to be duplicated.

rajkan01 replied:
Ideally, we have to include the source under test only that test covers and use dependency with stubs and all the config added in target_compile_definitions is normally generated by mbed configure command based on https://github.com/ARMmbed/mbed-os/blob/master/storage/filesystem/fat/ChaN/mbed_lib.json this configuration file.
In Unittest case we are not running any `mbedtools configure` command so link against this library is not helpful

Patater commented on storage/kvstore/filesystemstore/tests/UNITTESTS/FileSystemStore/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+        MBED_LFS_READ_SIZE=64
+        MBED_LFS_PROG_SIZE=64
+        MBED_LFS_BLOCK_SIZE=512
+        MBED_LFS_LOOKAHEAD=512
```
We could remove this config stuff if we link against a littlefs library target instead.

rajkan01 replied:
Can we address this issue as part of "Refactor components for independent unit testing"?

rwalton-arm commented on CMakeLists.txt:
```diff
@@ -5,6 +5,24 @@
+option(MBED_BUILD_UNITTESTS "Run unit tests only." OFF)
+
+if(MBED_BUILD_UNITTESTS)
+    project(mbed-unittests)
```
Should we be creating a project just for the unittests? Does it make sense to always create an mbed-os project instead? There are probably other cases where users may want to build from mbed-os root. It may also be useful to set the current mbed-os version so users can depend on specific versions of mbed.

rajkan01 replied:
As suggested we created a new mbed-os project

rwalton-arm commented on UNITTESTS/CMakeLists.txt:
```diff
@@ -1,240 +1,68 @@
+include(FetchContent)
+
+# Download and unpack googletest
+if(NOT GTEST_FOUND)
```
Is GTEST_FOUND ever true? I don't see where we include FindGtest.cmake, but maybe I missed it somewhere.

rajkan01 replied:
https://github.com/ARMmbed/mbed-os/pull/14426/commits/fb4998abfae82cab8be65f4b2e25dbe551a65cdf#r607783625

rajkan01 committed:
sort stubs into libraries

rajkan01 committed:
CMake: Update mbed os, unittest CMake and add CMAKE_CROSSCOMPILING guard

- Add a new MbedOS project in mbed os root CMake which can be used along with
  BUILD_TESTING conditional check for enabling the unittest build
- Update UNITTEST CMake for setting the CMake configuration like c, cxx flags etc.,
- Add if CMAKE_CROSSCOMPILING conditional check wherever target configuration check
  and toolchain configuration to avoid such configuration gets included for unittest build.

rajkan01 committed:
CMake: Add add_subdirectory of unittests

- add every libraries unittest directory into respective CMake
  which allows to include unittest source into build based on
  MBED_BUILD_UNITTESTS flag

rajkan01 committed:
CMake: Refactor drivers unittest cmake

rajkan01 committed:
CMake: Remove duplicate documents

- Remove the readme as it is a duplicate content of official document
  https://os.mbed.com/docs/mbed-os/v6.9/debug-test/unit-testing.html

rajkan01 committed:
CMake: Fix ChainingBlockDevice unittest

- ChainingBlockModuleTest test case compares two strings with EXPECT_EQ
which normally compare strings residing memory address so replaced it with EXPECT_STREQ  to compare strings.

rajkan01 committed:
CMake: Refactor platform unittest CMake

- Add CMake configuration file test suites

rajkan01 committed:
CMake: Refactor event unittest cmake

- Added CMake configuration file into events unittests

rajkan01 committed:
CMake: Refactore cellular framework unittest

- Add CMake configuration file into framework AT, device, common
unittest

rajkan01 committed:
CMake: Fix DTLS socket unittest

- Binding the socket address into network stack to avoid socket connect API call hangs in the unittest

LDong-Arm commented on UNITTESTS/stubs/CMakeLists.txt:
```diff
@@ -0,0 +1,148 @@
+# Copyright (c) 2021 ARM Limited. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+add_library(mbed-stubs-headers INTERFACE)
+add_library(mbed-headers INTERFACE)
+add_library(mbed-headers-base INTERFACE)
+add_library(mbed-headers-platform INTERFACE)
+add_library(mbed-headers-connectivity INTERFACE)
+add_library(mbed-headers-storage INTERFACE)
+add_library(mbed-headers-drivers INTERFACE)
+add_library(mbed-headers-hal INTERFACE)
+add_library(mbed-headers-events INTERFACE)
+add_library(mbed-headers-rtos INTERFACE)
```
I think the names are a bit confusing - they are for unit tests only.
What about just a single `mbed-stubs-headers`:
```cmake
add_library(mbed-stubs-headers INTERFACE)

# platform
target_include_directories(mbed-stubs-headers
    INTERFACE
        ...
)


# storage
target_include_directories(mbed-stubs-headers
    INTERFACE
        ...
)

...
```
as CMake allows us to call `target_include_directories()` as many times as we need on one library. It's a pretty neat aspect of CMake IMO.

LDong-Arm replied:
Or, a more major question: Is it possible for each test suite to directly link the original library (e.g. `mbed-storage`, `mbed-events`, etc.)? Any source files that require real Mbed targets can be excluded in a similar way to aad3974ad5ab5472f34643915e210d280715785e.

(Personally I'm open for discussion, as we also don't want to overcomplicate stuff and this PR is already in a working state.)

paul-szczepanek-arm replied:
Unit tests need to be able to include headers selectively, for example only storage headers - not as a single library that contains all of them. You can of course remove this from mbed-os. I will then have to include my headers by hand in my tests or recreate these libs outside of mbed-os.

LDong-Arm replied:
> Unit tests need to be able to include headers selectively, for example only storage headers

It's not currently the case in this PR - it consolidates all `mbed-header-<component>` into `mbed-headers`, and links every test to the latter.

paul-szczepanek-arm replied:
Nothing gets linked here, they are all interface libraries. The mbed-headers is for convenience: if you need all headers you just require it instead of listing all the individual ones. If you only want a subset you can require the individual components. Best of both worlds. My unit tests use that.

LDong-Arm replied:
> The mbed-headers is for convenience

This is what I meant by "link" (in CMake sense `target_link_libraries`), not linker sense.

I think it'd be ideal for tests to reuse the original library (e.g. mbed-storage, mbed-events, etc.) that Mbed targets use, rather than creating separate header libraries. But if that's not feasible, then the current approach is okay.

rajkan01 replied:
Including the original library will bring additional source c/cpp files into the build, ideally, unittests should list the required sources (unit under test and test code ) but I haven't tried to use direct original libraries.

LDong-Arm replied:
Okay the way it is, I think

LDong-Arm commented on UNITTESTS/stubs/connectivity/CMakeLists.txt:
```diff
@@ -0,0 +1,63 @@
+        MDMRXD=NC
+)
+
+target_sources(mbed-stubs-connectivity
+    PRIVATE
+        aes_stub.c
+        AT_CellularContext_stub.cpp
+        AT_CellularDevice_stub.cpp
```
Open question to other reviewer: now we have
```
UNITTESTS
  - stub
    - connectivity
    - drivers
    - platform
    - ...
```
Rather than duplicating the top-level mbed-os layout, what about moving stubs into their respective components? Any preferences?

rwalton-arm replied:
> Open question to other reviewer: now we have
> 
> ```
> UNITTESTS
>   - stub
>     - connectivity
>     - drivers
>     - platform
>     - ...
> ```
> 
> Rather than duplicating the top-level mbed-os layout, what about moving stubs into their respective components? Any preferences?

I would also prefer it if the stubs were moved into their respective components. I don't really see the value in duplicating the mbed-os layout here either.

rajkan01 replied:
As discussed, we will plan this work as part of `Refactor components for independent unit testing`

LDong-Arm commented on UNITTESTS/mbed_unittest.py:
```diff
@@ -71,9 +71,8 @@ def _mbed_unittest_test(options, cwd, pwd):
-        src_path = os.path.relpath(pwd, options.build)
-        tool.create_makefiles(path_to_src=src_path,
+        # Create makefiles        
+        tool.create_makefiles(path_to_src=cwd,
```
As we're now using "modern CMake" for both Mbed CLI 1 & Mbed CLI 2, we can safely delete the old-fashion `unittest.cmake`s from the entire Mbed OS:
```bash
find . -name "unittest.cmake" -exec rm {} \;
```
I've tested it. This is a clean up we need to do.

paul-szczepanek-arm replied:
I agree, we can safely remove these - I think these were removed in the original PR before we started this second attempt.

rajkan01 replied:
Removing the unittest.cmake in the separate PR https://github.com/ARMmbed/mbed-os/pull/14547

LDong-Arm commented on CMakeLists.txt:
```diff
@@ -26,80 +42,83 @@ target_link_libraries(mbed-baremetal
-                " we are using the standard C library instead."
-            )
-            set(MBED_C_LIB "std" CACHE STRING "")
+if(${CMAKE_CROSSCOMPILING})
```
Well, this is kind of okay.

But I think `MBED_TARGET` might be more direct because it's generated by mbed-tools. If this variable is not set, then it's simply not an Mbed target. What do others think?

rajkan01 replied:
As discussed, there is a possibility of `mbed_config.cmake` generated by the application and it can also build the unittest so we go with `CMAKE_CROSSCOMPILING` flag checking.

LDong-Arm reviewed (changes requested):
Thanks for the big PR. I've added a few comments regarding the overall implementation.

rwalton-arm commented on UNITTESTS/stubs/storage/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+        ExhaustibleBlockDevice_stub.cpp
+        FlashSimBlockDevice_stub.cpp
+        HeapBlockDevice_stub.cpp
+        #MBRBlockDevice_stub.cpp
```
Should we remove this commented line?

rajkan01 replied:
I have enabled this commented line again and fixed the building failure of the missing definition of `MBED_PACKED`

rwalton-arm commented on UNITTESTS/target_h/cmsis_os.h:
```diff
@@ -27,4 +27,5 @@
 typedef struct {
 } osEvent;
 
+typedef int32_t osStatus;
```
Should this change be in a separate commit? Is this change needed to "sort stubs into libraries"?

rajkan01 replied:
yes this change required to sort headers stubs into libraries and better to keep in the same commit

rwalton-arm commented on UNITTESTS/stubs/connectivity/CMakeLists.txt:
```diff
@@ -0,0 +1,63 @@
+
+target_link_libraries(mbed-stubs-connectivity
+    PRIVATE
+        mbed-headers
```
Aren't we adding too much here? Do we need to add "all available headers" to every stub library?

rajkan01 replied:
For connectivity stubs library sources use headers from the below list of headers library except `mbed-headers-storage` so we added mbed-headers to avoid manually include every headers library individually I think it is the same for other stubs library
```
mbed-os/UNITTESTS/stubs/CMakeLists.txt:

target_link_libraries(mbed-headers
    INTERFACE
        mbed-headers-base
        mbed-headers-platform
        mbed-headers-connectivity
        mbed-headers-storage
        mbed-headers-drivers
        mbed-headers-hal
        mbed-headers-events
        mbed-headers-rtos
)
```

rwalton-arm commented on storage/kvstore/filesystemstore/tests/UNITTESTS/FileSystemStore/CMakeLists.txt:
```diff
@@ -0,0 +1,48 @@
+
+target_include_directories(${TEST_NAME}
+    PRIVATE
+        ${MBED_PATH}
```
Aren't we removing MBED_PATH? Do we need to add `mbed-os` as an include directory for this test?

rajkan01 replied:
This is a problem with LittleFileSystem.h and LittleFileSystem.cpp as it refers to the littflefs headers with the complete path like `storage/filesystem/littlefs/littlefs/lfs_util.h` and fixed the path issue to avoid including the mbed-os into include directories

rwalton-arm commented on storage/blockdevice/tests/UNITTESTS/blockdevice/CMakeLists.txt:
```diff
@@ -0,0 +1,17 @@
+    ExhaustibleBlockDevice
+)
+
+mbed_add_all_subdirectories("${MBED_ADD_UNITTEST_SUBDIRS}")
```
What's the point of this function? It's marginally less typing than just calling the regular `add_subdirectory` and it adds an extra level of indirection which impedes readability.

rajkan01 replied:
I replaced `mbed_add_all_subdirectories` with direct `add_subdirectory`

rwalton-arm commented on storage/blockdevice/tests/UNITTESTS/blockdevice/FlashSimBlockDevice/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+    PRIVATE
+        mbed-headers
+        mbed-stubs-headers
+        mbed-stubs-platform      
```
Doesn't `mbed-stubs-platform` already depend on `mbed-stubs-headers`? Not sure why we also depend on it again here.

rajkan01 replied:
`mbed-stubs-platform` is target_link_libraries with PRIVATE against `mbed-stubs-headers` and it will not visible in FlashSimBlockDevice CMake and based on our previous discussion here https://github.com/ARMmbed/mbed-os/pull/14285#discussion_r576287699 decided to go with PRIVATE

rwalton-arm replied:
I think we could've avoided this circular flow of dependencies if we just linked to the header libraries being used in the stubs instead of adding "all available headers" everywhere, as I mentioned in another comment. This is probably ok for now though.

rwalton-arm commented on connectivity/cellular/tests/UNITTESTS/framework/AT/at_cellularstack/CMakeLists.txt:
```diff
@@ -0,0 +1,37 @@
+
+target_sources(${TEST_NAME}
+    PRIVATE
+        ${MBED_PATH}/connectivity/cellular/source/framework/AT/AT_CellularStack.cpp
```
More MBED_PATH...

rajkan01 replied:
Fixed

LDong-Arm commented on CMakeLists.txt:
```diff
@@ -127,24 +143,38 @@ add_subdirectory(drivers)
+    add_subdirectory(targets)
+endif()
+
+if(${MBED_BUILD_UNITTESTS})
```
Could directly check `NOT ${CMAKE_CROSSCOMPILING}`?

rajkan01 replied:
It is a good point @LDong-Arm, I will check

rajkan01 replied:
Refactored this source code

LDong-Arm commented on UNITTESTS/unit_test/test.py:
```diff
@@ -86,6 +86,7 @@ def create_makefiles(self,
         args = [cmake,
                 "-G",
                 generator,
+                "-DBUILD_TESTING=ON"
```
`include(CTest)` sets this automatically. As we're free to run any commands here, why not run the same `ctest` command that we use with Mbed CLI 2?

rajkan01 replied:
I would like to avoid major rework on this script as we are proposing to use CMake/CTest command for unittest from Mbed CLI 2 and deprecate this python module scripts along with Mbed CLI 1 and time being these changes to make backward compatibility with Mbed CLI 1

LDong-Arm commented on CMakeLists.txt:
```diff
@@ -15,6 +17,17 @@ list(APPEND CMAKE_MODULE_PATH
     "${mbed-os_SOURCE_DIR}/platform/FEATURE_EXPERIMENTAL_API/FEATURE_PSA/TARGET_TFM/TARGET_TFM_LATEST/scripts;${mbed-os_SOURCE_DIR}/targets/TARGET_Cypress/scripts;${mbed-os_SOURCE_DIR}/targets/TARGET_NXP/scripts"
 )
 
+option(BUILD_TESTING "Run unit tests only." OFF)
```
This is set by CTest, see my other comment

rajkan01 replied:
We want to get the `BUILD_TESTING` argument from the cmake/ctest command line and manually adding option with OFF helps to override `BUILD_TESTING` gets `ON` by `include(CTest)` or `enable_testing`

LDong-Arm reviewed (commented):
The old CMake definitions for unit tests work with fairly old version of python. So this change means that even if a user sticks to the old Mbed CLI 1 (mbed-cli), they will also need to install CMake >= 3.19 to be able to run unit tests? This is worth documenting, at least in "Impact of changes" in the PR description.

Otherwise the overall structure of this PR looks good to me, as further refactoring will be done separately.

rajkan01 committed:
CMake: Run the unit tests in travis CI

rajkan01 committed:
CMake: fix lifflefs filesystem header include path

- LittleFileSystem.h and LittleFileSystem.cpp refers to the littflefs headers with the complete path like storage/filesystem/littlefs/littlefs/lfs_util.h and fixed the path issue

LDong-Arm reviewed (approved):
Looks good to me. Given the scope of this PR, wait for others' re-review.

wernerlewis commented on connectivity/netsocket/tests/UNITTESTS/netsocket/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+add_subdirectory(TLSSocket)
+add_subdirectory(TLSSocketWrapper)
+add_subdirectory(UDPSocket)
+add_subdirectory(WiFiAccessPoint)
```
Missing a newline here.

wernerlewis commented on connectivity/cellular/tests/UNITTESTS/framework/AT/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+add_subdirectory(at_cellularinformation)
+add_subdirectory(at_cellularnetwork)
+add_subdirectory(at_cellularsms)
+add_subdirectory(at_cellularstack)
```
Missing a newline here.

wernerlewis reviewed (commented):
There are a couple of missing newlines, and some of the lines in commit messages are very long, it would help to wrap those. Other than that, looks good to me.

rajkan01 committed:
Fix missing newline

rwalton-arm commented on UNITTESTS/stubs/hal/CMakeLists.txt:
```diff
@@ -0,0 +1,25 @@
+    PRIVATE
+        mbed-headers
+        mbed-stubs-headers
+        gcov
```
This won't work with AppleClang. We need to pass `target_link_options(mbed-stubs-hal --coverage)` instead. That will alias to the correct linker argument on gcc and AppleClang. The current setup causes a build failure with the error `library not found for -lgcov` when building with AppleClang.

rwalton-arm commented on UNITTESTS/stubs/platform/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+    PRIVATE
+        mbed-headers
+        mbed-stubs-headers
+        gcov
```
As above, this won't work for AppleClang. We need to pass target_link_options(mbed-stubs-platform --coverage) instead. That will alias to the correct linker argument on gcc and AppleClang. The current setup causes a build failure with the error `library not found for -lgcov` when building with AppleClang.

rajkan01 committed:
CMake: Update hal and platform stub CMake linker options for coverage

 - Armclang is not working when we add gcov as target_link_libraries,
modified to pass as linker options with --coverage so it can work armclang and gcc

0xc0170 reviewed (dismissed):
Do not forget to add a note to our docs about CMAKE_CROSSCOMPILING. I'll review docs part of   this change , it should be there (not directly connected to unittests, they only benefit from it, but in CMake docs - as if anyone adds new CMakeLists, they should be aware of it and how to use it).

donatieng commented on CMakeLists.txt:
```diff
@@ -135,24 +147,35 @@ add_subdirectory(drivers)
 add_subdirectory(rtos)
-add_subdirectory(targets)
+
+if(${CMAKE_CROSSCOMPILING})
```
Do we really need to add all those `CMAKE_CROSSCOMPILING` if checks? Looks a bit messy - should this kind of special handling be kept within the `targets/` folder?

rajkan01 replied:
Yes to keep in the root directory for the below reasons:
- Added `CMAKE_CROSSCOMPILING` check is at the root CMakeLists.txt to provide better visibility of what is getting included/excluded to the build by default.
- The `targets` directory sources are not having any unit tests so it is easy to identify which are the sources getting added for unit test build

donatieng replied:
I understand what you're trying to achieve but actually, the more I look at it, the more I'm worried about the proposed solution 😅. The root `CMakeLists.txt` file should be kept as simple as possible, and unit testing build logic should be with each component. For instance, it's true that today `targets/` don't have any unit tests, but I would argue that we shouldn't discourage that (quite the opposite).

rajkan01 replied:
@donatieng I have moved the `CMAKE_CROSSCOMPILING` guard inside the library (connectivity, storage, events, targets) CMake, please re-review

donatieng commented on CMakeLists.txt:
```diff
@@ -135,24 +147,35 @@ add_subdirectory(drivers)
+if(${CMAKE_CROSSCOMPILING})
+    add_subdirectory(targets)
+    # The directories below contain optional target libraries
+    add_subdirectory(connectivity EXCLUDE_FROM_ALL)
```
Same, I understand why you're doing that but it's adding a bit of un-necessary complexity, as we should assume that most code we maintain will eventually have unit tests. Any reason why we can't remove `EXCLUDE_FROM_ALL` in any case?

rajkan01 replied:
Yes most of the components have unit test and we want to include to the build when we invoke unit test cmake command but in case of application build (Cross compiling), we would like to exclude to the build by default.  When the application cmake target_link_libraries to any of the excluded component then it gets added into the build.

donatieng replied:
I see why but this has implications beyond unit tests. I would suggest the specific unit testing targets are excluded from the `all` target in the appropriate places, not at the top level here.

rwalton-arm replied:
If we remove `EXCLUDE_FROM_ALL` then all the added CMake targets will be compiled when we try to build the CMake `all` target (mbed-tools tries to build the all target as it has no knowledge of CMake target names). Using `EXCLUDE_FROM_ALL` here allows us to specify which components of Mbed are built by using target_link_libraries in the consuming application. This isn't an ideal set-up but it was a pre-existing issue. I'd suggest we add an issue to the CMake milestone to tidy up here.

0xc0170 replied:
> I'd suggest we add an issue to the CMake milestone to tidy up here.

@rwalton-arm can you create an issue please for this?

donatieng commented on UNITTESTS/mbed_unittest.py:
```diff
@@ -71,9 +71,8 @@ def _mbed_unittest_test(options, cwd, pwd):
-        src_path = os.path.relpath(pwd, options.build)
-        tool.create_makefiles(path_to_src=src_path,
+        # Create makefiles        
+        tool.create_makefiles(path_to_src=cwd,
```
Why is this changed?

rajkan01 replied:
The previous unit test build is also based on CMake and the build starts from `mbed-os/UNITTESTS` directory (Present working directory) but with the new changes build starts from the mbed-os root directory (Current working directory) so backward compatibility to support mbed CLI 1 just tweak here to start the build from the current working directory (mbed-os root dir) and we may remove this mbed-os/UNITTESTS/*.py module once deprecate CLI1 as we can able add this Cmake command calling directly into mbed CLI2

donatieng replied:
Cool, thanks for the explanation!

donatieng commented on UNITTESTS/stubs/CMakeLists.txt:
```diff
@@ -0,0 +1,149 @@
+        mbed-headers-rtos
+)
+
+target_include_directories(mbed-headers-platform
```
This feels like duplication to me - as these are "real" headers, can't we define them directly in the relevant directory's `CMakeLists.txt` (for instance `/platform/CMakeLists.txt`? This goes quite against any future componentisation effort. The "real" targets can then link to these "headers" targets.

rajkan01 replied:
The upcoming task in the backlog to create Refactor components for independent unit testing as part of that we will do the further cleanup in `UNITTESTS/stubs`, `UNITTESTS/target_h` to move this to the respective components directory. For example, we will move the platform stubs and header library creation inside `mbed-os/platform/tests/UNITTESTS` directory

donatieng replied:
Any reason why that work is not made as part of this PR?

rajkan01 replied:
As the first set of work planned to modify existing unit test CMake support to proper CMake way of doing things and further next phase of work will plan moving stubs/target_h closure to library

0xc0170 replied:
@rajkan01 Can you create tracking issue for this planned work?

rajkan01 replied:
> 
> 
> @rajkan01 Can you create tracking issue for this planned work?

@0xc0170 Already raised Github issues for this work

donatieng reviewed (dismissed):
Thanks @rajkan01! Lots of good stuff here.

One slight concern from my standpoint: stubs and headers are defined "globally" across the OS, as opposed to the relevant component directory. The risk is that it becomes harder to maintain these components (as you have to change the stubs which are in a different place - which is not obvious, and there is duplication in terms of listing the relevant header files for instance).

Otherwise I see you've dropped the documentation, would be good to add something back except if I've missed something else.

Finally, more of a question - what's the impact (if any) on CLI1?

rajkan01 committed:
CMake: Move CMAKE_CROSSCOMPILING check inside library CMake

donatieng reviewed (approved):
Thanks for the additional comments and rework @rajkan01 and @rwalton-arm.
Let's get this in as it's a good starting point, but please raise issues for what needs to happen next :). Great work!
